### DISCUSSION TITLE
Refactoring Navigation

### DISCUSSION THREAD

**Discussion by stellanw** [neuefische/questions#821] [Web] [answered]
Refactoring Navigation

Hello,
I am currently working on our Navigation component. It's used dynamically on different pages. The main function "create" is implemented on main page ("/"). And it's connected with an overlay which opens the form and hide the navigation. 
But to implement the same behavior to the joined and calendar page, I would have to copy paste the "handleCreateClick" and "handleCloseClick". I could uplift the isCreateMode useState to global but thats all. 

Right now the status is that the functionality of the create button works at main page, but not at joined or calendar page.

Maybe Someone has a brilliant idea.

Here is my current status. [It is accessible at this link.](https://github.com/alex96dev/Easy-Meetup-with-Friends) 

The Part of Code I'm talking about:
`      {!isCreateMode && <Navigation onCreateClick={handleCreateClick} />}
      {isCreateMode && (
        <Overlay>
          <CardForm
            pageTitle="Create your activity!"
            onCancel={handleCloseClick}
            setIsCreateMode={setIsCreateMode}
            setIsEditMode={setIsEditMode}
            isEditMode={false}
            onSubmit={onSubmit}
          />
        </Overlay>
      )}
`
Thank you for your help!

**Comment by Kheirah** [comment, marked as answer]
Without using the Context API or external libraries like Zustand/Redux, you can implement a higher order component which manages the state of the component that is passed as a prop.

### Code up a higher order component

```js
// withVisibility.js
import React, { useState } from 'react';

const withVisibility = (WrappedComponent) => {
    return (props) => {
        const [isVisible, setIsVisible] = useState(false);

        const toggleVisibility = () => setIsVisible(!isVisible);

        return <WrappedComponent 
                 isVisible={isVisible} 
                 toggleVisibility={toggleVisibility}
                 {...props} />;
    };
};

export default withVisibility;
```

### Using the higher order component

`MyComponent` would be the `Navigation` component in your case.

```js
import React from 'react';
import withVisibility from './withVisibility';

const MyComponent = ({ isVisible, toggleVisibility }) => (
    <>
        {isVisible && <div>This component is now visible!</div>}
        <button onClick={toggleVisibility}>Toggle Visibility</button>
    </>
);

export default withVisibility(MyComponent);
```

Take it with a grain of salt; the code is exemplary and you need to adjust it to fit your use case.

**Reply by stellanw** [reply]
Thanks! Zustand was a very good hint. I used the Zustand library and it works great for what I needed.

```javascript
import create from "zustand";

const useStore = create((set) => ({
  isCreateMode: false,
  setIsCreateMode: (value) => set({ isCreateMode: value }),
  isEditMode: false,
  setIsEditMode: (value) => set({ isEditMode: value }),
  handleCreateClick: () => set({ isCreateMode: true }),
  handleCloseClick: () => set({ isCreateMode: false }),
  handleEditClick: () => set({ isEditMode: true }),
}));

export default useStore;

```
